File: SpringBootAngularServer/src/main/java/com/ComputerPartChecker/server/service/BaseService.java
package com.ComputerPartChecker.server.service;

import com.ComputerPartChecker.server.repository.BaseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public abstract class BaseService<E, K, R extends BaseRepository<E, K>> {
    protected R repository;

    @Autowired
    public BaseService(R repository) {
        this.repository = repository;
    }

    @Autowired
    public final void setRepository(R repository) {
        this.repository = repository;
    }

    public List<E> findAll() {
        return repository.findAll();
    }

    public E add(E entity) {
        return repository.save(entity);
    }

    public void delete(E entity) {
        repository.delete(entity);
    }

    public E edit(E entity) {
        E persistanceEntity = repository.save(entity);
        repository.refresh(persistanceEntity);
        return persistanceEntity;
    }
}

File: SpringBootAngularUI/src/app/service/base.service.ts
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export abstract class BaseService<E> {

  protected url: string;
  private server: string = 'http://localhost:8080';

  constructor(protected http: HttpClient, url: string) {
    this.url = this.server + url;
  }

  public findAll(): Observable<E[]> {
    return this.http.get<E[]>(this.url + '/findAll');
  }

  public add(entity: E) {
    return this.http.post<E>(this.url + '/add', entity);
  }

  public delete(entity: E) {
    return this.http.post<E>(this.url + '/delete', entity);
  }

  public edit(entity: E) {
    return this.http.post<E>(this.url + '/edit', entity);
  }
}
